cou/steps/analyze.py

```diff
@@ -0,0 +1,205 @@
+# Copyright 2023 Canonical Limited.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Functions for analyzing an OpenStack cloud before an upgrade."""
+from __future__ import annotations
+
+import logging
+from collections import defaultdict
+from dataclasses import dataclass, field
+from typing import Any, Dict, Iterable, Union
+
+import yaml
+from juju.client._definitions import ApplicationStatus
+
+from cou.exceptions import CommandRunFailed
+from cou.utils.juju_utils import (
+    async_get_application_config,
+    async_get_full_juju_status,
+    async_run_on_unit,
+)
+from cou.utils.openstack import CHARM_TYPES, get_os_code_info
+from cou.utils.upgrade_utils import extract_charm_name_from_url
+
+
+@dataclass
+class Analysis:
```

Comment:
it seems we're missing a method to determine the overall version of the cloud, though it doesn't have to be part of this changeset

Reply:
For the Analysis part it is not viable to to determine the openstack version. Because applications(even units) can have different versions. We will consider these differences in the planning part and create an upgrade plan according to these versions.

Reply:
I agree with Erhan here.

Reply:
I imagine either works, though I see determining what kind of cloud you're dealing with as a starting point (ussuri? victoria? partially-upgraded wallaby?) more in the domain of the analysis part rather than in the plan one